libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/RoutesUpdateReasonHelper.kt

```diff
@@ -0,0 +1,115 @@
+import com.mapbox.navigation.utils.internal.logD
+
+internal class RoutesUpdateReasonHelper(mapboxNavigation: MapboxNavigation) {
+
+    private var offRoute = false
+
+    private val offRouteObserver = OffRouteObserver { offRoute ->
+        this.offRoute = offRoute
+    }
+
+    private companion object {
+        private val TAG = Tag("MbxRouteChangeReasonHelper")
+    }
+
+    init {
+        mapboxNavigation.registerOffRouteObserver(offRouteObserver)
+    }
+
+    /**
+     * Provides the reason why the routes have been updated.
+     *
+     * Reason is provided based on the following (priority matches with sequence):
+     * - if [newRoutes] is empty -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_CLEAN_UP];
+     * - if [legacyRoutes] is empty and [newRoutes] is NOT empty -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES];
+     * - if off-route event happened and [RouteOptions.coordinatesList] are same for primary routes of
+     * [legacyRoutes] and [newRoutes] (except **origin** coordinate) -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES];
+     * - if [DirectionsRoute.requestUuid] (see [isSameUuid]) and [DirectionsRoute.geometry]
+     * (or **steps' names**, see [isSameRoute]) are the same of primary routes
+     * of [legacyRoutes] and [newRoutes] (except **origin** coordinate) -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES];
+     * - if [RouteOptions.coordinatesList] are same for primary routes of [legacyRoutes] and
+     * [newRoutes] (except **origin** coordinate) -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_ALTERNATIVE_ROUTES];
+     * - otherwise -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES].
+     *
+     * Note: function must be invoked before set a new route to NN, because it relies on
+     * off-route events.
+     */
+    @RoutesExtra.RoutesUpdateReason
+    internal fun getReason(
+        legacyRoutes: List<DirectionsRoute>,
+        newRoutes: List<DirectionsRoute>
+    ): String {
+        when {
+            newRoutes.isEmpty() ->
+                return RoutesExtra.ROUTES_UPDATE_REASON_CLEAN_UP
+            legacyRoutes.isEmpty() && newRoutes.isNotEmpty() ->
+                return RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES
+        }
+
+        val isSameRoutes = sameRoutes(legacyRoutes, newRoutes)
+        val isSameUuid = sameUuid(legacyRoutes, newRoutes)
+        val isSameRouteOptionsCoordinates = sameRouteOptionsCoordinates(legacyRoutes, newRoutes)
+        return when {
+            offRoute && isSameRouteOptionsCoordinates ->
+                RoutesExtra.ROUTES_UPDATE_REASON_REROUTE_ROUTES
+            isSameUuid && isSameRoutes ->
+                RoutesExtra.ROUTES_UPDATE_REASON_REFRESH_ROUTES
+            isSameRouteOptionsCoordinates ->
+                RoutesExtra.ROUTES_UPDATE_REASON_ALTERNATIVE_ROUTES
+            else -> RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES
+        }
+    }
+
+    private fun sameUuid(
```

Comment:
```suggestion
    private fun sameUuidForPrimaryRoute(
```

Same as above, it took me some time to notice this method only compares the primary route.